### DISCUSSION TITLE
Blocking? start_polling and idle

### DISCUSSION THREAD

**Discussion by gclintj** [python-telegram-bot/python-telegram-bot#2623] [Q&A] [answered]
Just got started by reading echobot.py

> start_polling() is non-blocking and will stop the bot gracefully.

What does it mean by blocking and non-blocking here?
How would start_polling stop the bot and why is it gracefully?
Thanks!

**Comment by harshil21** [comment, marked as answer]
> What does it mean by blocking and non-blocking here?

Blocking means that your code will block further execution until that operation is done. An example would be an infinite while loop. The code after the while loop is never run. Non-blocking would be the opposite. That is usually done with threads. So the code after `start_polling()` runs.

> How would `start_polling` stop the bot and why is it gracefully?

The bot stops running when it receives a [signal](http://www.cs.kent.edu/~ruttan/sysprog/lectures/signals.html). When this happens, threads & connection pools are closed gracefully, data is flushed to files (if you enabled persistence), and optionally if you've passed the param `user_sig_handler` in [Updater](https://python-telegram-bot.readthedocs.io/en/stable/telegram.ext.updater.html), then that is run too.

**Reply by gclintj** [reply]
Thanks. I feel I have a better understanding of start_polling(), idle(), and the term blocking.

Just to clarify, by
```
    # Run the bot until you press Ctrl-C or the process receives SIGINT,
    # SIGTERM or SIGABRT. This should be used most of the time, since
    # start_polling() is non-blocking and will stop the bot gracefully.
    updater.idle()
```
It means 
**using idle()** will stop the bot gracefully, or 
**start_polling()** will stop the bot gracefully?

Thanks!

**Reply by harshil21** [reply]
`idle()` will stop the bot gracefully

**Reply by gclintj** [reply]
Thanks!
I would suggest editing the comments mentioned for clarification.
